File: 0016-3sum-closest/0016-3sum-closest.py
class Solution(object):
    def threeSumClosest(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: int
        """
        max_diff=float('inf')
        i=0
        res_sum=0
        nums.sort()
        while i<len(nums)-1:
            left=i+1
            right=len(nums)-1
            while left<right:
                result=nums[i]+nums[left]+nums[right]
                diff=abs(result-target)
                if diff<max_diff:
                    max_diff=diff
                    res_sum=result
                if result<target:
                    left+=1
                elif result>target:
                    right-=1
                else:
                    break
            i+=1
        return res_sum

File: 0016-3sum-closest/0016-3sum-closest.java
class Solution {
    public int threeSumClosest(int[] nums, int target) {
        //brute force;
        // int ans= Integer.MAX_VALUE;
        // int result=0;
        // for(int i=0;i<nums.length;i++){
        //     for(int j=i+1;j<nums.length;j++){
        //         for(int k=j+1;k<nums.length;k++){
        //             int sum=nums[i]+nums[j]+nums[k];
        //             int diff=Math.abs(target-sum);
        //             if(diff<ans){
        //                 ans=diff;
        //                 result=sum;
        //             } 
        //         }
        //     }
        // }
        // return result;
        int ans=Integer.MAX_VALUE;
        int result=0;
        Arrays.sort(nums);
        for(int i=0;i<nums.length;i++){
            int j=i+1;
            int k=nums.length-1;
            while(j<k){
                int sum= nums[i]+nums[j]+nums[k];   
                int diff= Math.abs(sum-target);
                if(diff<ans){
                    ans=diff;
                    result=sum;
                }
                if(sum>target){
                    k-=1;
                }else{
                    j+=1;
                }
            }
        }
        return result;
    }
}